brave/src/test/java/brave/features/handler/FinishedChildrenHandler.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.features.handler;
+
+import brave.handler.FinishedSpanHandler;
+import brave.handler.MutableSpan;
+import brave.propagation.TraceContext;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+import static java.util.Collections.emptyList;
+
+public abstract class FinishedChildrenHandler extends FinishedSpanHandler {
+
+  protected abstract void handle(MutableSpan parent, Iterable<MutableSpan> children);
+
+  /** This holds the children of the current parent until the former is finished or abandoned. */
+  final ParentToChildren parentToChildren = new ParentToChildren();
+
+  @Override public boolean supportsOrphans() {
+    return true; // We need to clear the map entry even if the parent doesn't finish normally
+  }
+
+  @Override public boolean handle(TraceContext context, MutableSpan span) {
+    if (!context.isLocalRoot()) { // a child
+      parentToChildren.add(context.parentIdString(), span);
+    }
+
+    Iterable<MutableSpan> children = parentToChildren.remove(context.spanIdString());
```

Comment:
Wonder if we should go ahead and copy into another list when passing to handler.

Reply:
I guess I'm confusing things by being inconsistent. I intentionally used synchronized above because we don't have a better collection type at the moment. That isn't not good for performance, so me avoiding array allocation here is probably not intuitive.

In real code I'd probably not do that, as it is possibly going to happen inline with a random call to Span.finish (time is stolen to invoke this hook). I'd probably prefer exposing Iterator instead..

Reply:
updated with some notes on why.. ack allocation defence it is unimportant with how it is currently implemented though.